File: patterns/rule-refresh/rules-python/rule_refresh.py
from openhab import rule, Registry
from openhab.triggers import ItemCommandTrigger, ItemStateChangeTrigger


def build_triggers_from_metadata():
    # Beispiel: alle Items mit dem Metadaten-Namespace "triggerRule" einsammeln
    triggers = []
    for item in Registry.getItems():
        if item.getMetadata().get("triggerRule") is not None:
            triggers.append(ItemStateChangeTrigger(item.name))
    return triggers


@rule(name="Dynamische Metadaten-Regel")
class DynamischeMetadatenRegel:
    def buildTriggers(self):
        found = build_triggers_from_metadata()
        if not found:
            self.logger.warn("Keine passenden Items gefunden")
        return found

    def execute(self, module, input):
        self.logger.info(str(input.get("itemName")) + " hat sich geaendert (dynamischer Trigger)")


@rule(triggers=[ItemCommandTrigger("Reload_Item", "ON")])
class ReloadDynamischeRegel:
    def execute(self, module, input):
        # Ein erneuter Aufruf von buildTriggers() erfolgt automatisch,
        # sobald das Skript neu geladen wird (z. B. durch Speichern der Datei).
        self.logger.info("Trigger werden beim naechsten Neuladen des Skripts aktualisiert")

File: patterns/rule-refresh/rules-js/rule-refresh.js
const DYNAMIC_RULE_ID = "dynamic_metadata_rule";

function buildTriggersFromMetadata() {
  // Beispiel: alle Items mit dem Metadaten-Namespace "triggerRule" einsammeln
  return items.getItems()
    .filter((i) => i.getMetadata("triggerRule") !== null)
    .map((i) => triggers.ItemStateChangeTrigger(i.name));
}

function createDynamicRule() {
  const dynamicTriggers = buildTriggersFromMetadata();
  if (dynamicTriggers.length === 0) {
    console.warn("Keine passenden Items gefunden, Regel wird nicht erstellt");
    return;
  }

  rules.JSRule({
    id: DYNAMIC_RULE_ID,
    name: "Dynamische Metadaten-Regel",
    triggers: dynamicTriggers,
    overwrite: true,
    execute: (event) => {
      console.log(event.itemName + " hat sich geaendert (dynamischer Trigger)");
    }
  });
}

rules.JSRule({
  name: "Reload dynamische Regel",
  triggers: [triggers.ItemCommandTrigger("Reload_Item", "ON")],
  execute: (event) => {
    createDynamicRule();
  }
});

// Beim ersten Laden des Skripts einmal ausfuehren
createDynamicRule();
